src/moneypenny/kubernetes.py

```diff
@@ -1,165 +1,315 @@
+        except ApiException as e:
+            logger.exception("Exception creating pod")
+            raise K8sApiException(e)
+        logger.debug(f"Pod for {username} created: {status}")
 
-    def _pod_status(self, name: str) -> str:
-        """Return the status of the pod with the given name.
+    def delete_objects(self, username: str) -> None:
+        """Delete both the pod and its associated configmap, given a
+        username.
 
         Parameters
         ----------
-        name: Name of the pod to check on.
-
-        Returns
-        -------
-        The pod status: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#podstatus-v1-core
+        name: Username for the pod and configmap to delete.
 
+        Raises
+        ------
+        K8sApiException if the deletion failed.
         """
-        try:
-            logger.info(f"Checking status for pod {name}")
-            pod = self.apps_api.read_namespaced_pod(
-                name, self.namespace
-            )
-        except ApiException as exc:
-            logger.exception("Exception reading pod")
-            raise K8sApiException(exc)
-        return pod.status
+        self._pod_delete(username)
+        self._configmap_delete(username)
 
-    def pod_completed(self, name: str) -> bool:
+    def check_pod_completed(self, username: str) -> bool:
         """Return true if the pod completed successfully, false if it
         is pending or running, and raise an exception if any part of the
         execution failed.
-        
+
         Parameters
         ----------
-        name: Name of the pod to check on.
+        name: Username for the pod to check on.
 
         Returns
         -------
         True for successful completion, False for still-in-progress.
 
         Raises
         ------
-        OperationFailed if the pod encountered an error.
+        PodNotFound if the pod isn't there at all, OperationFailed if the pod
+        encountered an error, K8sApiException for some other Kubernetes error.
         """
-
+        pname = _name_object(username, "pod")
         try:
-            status=self._pod_status(name)
+            pod = self.api.read_namespaced_pod(pname, self.namespace)
+            status = pod.status
         except ApiException as exc:
             if exc.status == 404:
-                raise PodNotFound(f"Pod {name} not found")
+                raise PodNotFound(f"Pod {pname} not found")
+            logger.exception("Error checking on pod completion")
             raise K8sApiException(exc)
-        phase = status.phase
+        phase: str = status.phase
         if phase == "Succeeded":
             return True
         if phase == "Pending" or "Running":
             return False
         if phase == "Unknown":
-            raise OperationFailed(f"Pod {name} in Unknown phase")
+            raise OperationFailed(f"Pod {pname} in Unknown phase")
         raise OperationFailed(status.message)
-        
+
+    def _make_pod(
```

Comment:
Okay, after looking at this, and thinking about it a bit, I'm not sure I like this pattern.  First, there's a lot of functions to build it up, but I think they all just call each other once, so it's kind of building up the data, but it's more tricky for me to think about.  Maybe it is better to just have this as a yaml file that is templated.  Because what if we need another type, like a secret or a different configmap?  You can checkout the nublado2 part about just making objects from arbitrary yaml.  We could just have a convention that says there's a pod named "foo" and we wait for it to finish, using the code you've got here to check

Reply:
This might mean that we want to isolate these things in different namespaces though maybe... just to be able to delete everything in one go by cleaning up the namespace, so that might be a tricky change.  With this you have to kind of keep track of the cases where things exist.

Reply:
I've made all the containers just use yaml.  I was running this all in the Moneypenny namespace, because I don't think we need to give Moneypenny cluster privs and at least so far the provisioning/deprovisioning is namespace-independent.  There aren't that many things created: one pod and one configmap per user.